File: loop_rooms/loop_rooms.cpp
//  ECE NTUA
//  ProglangI
//  2021
//  Stulianos Kandylakis el17088
//  Kitsos Orfanopoulos el17025

#include <iostream>
#include <vector>
#include <algorithm>
#include <fstream>


using namespace std;


/* ----------------------- MAIN IDEA -------------------------------------------

We create a N x M dynamic array, called maze_doors. Each cell is marked with UDRL (up, down, right, left).
We read the given file (from arguments) and fill that maze..

Then we create a N x M dynamic array, called maze_escapes, with initial values '0' to each cell.

We run a DFS to find escapable rooms. If we find a visited one then the whole
path is "else", else is "good".
We use an auilary recursive function that follows the path of every cell.
We check whether the next room is already "visited", "bad" or "good".
We mark "visited" rooms with 2, "bad" with 3, "good" with 1.
Bad or visited are actually the same (see version of ML or Python).
--------------------------------------------------------------------------------
*/


int check_path(char ** doors, int **escapes, int i, int j, int N, int M){
  if ((escapes[i][j]==2) || (escapes[i][j]==3)) { // bad path (cycle)
    escapes[i][j]=3;
    return 3;
  }
  if (escapes[i][j]==1) { // good path
    return 1;
  }

  escapes[i][j] = 2; //we have visited this room

  char door = doors[i][j];

  switch(door) {
    //----------- left door -------------
      case 'L':
        if (j==0)  {
          escapes[i][j] = 1;
          return 1;
        }
        else{
            escapes[i][j] = check_path(doors, escapes, i, j-1, N, M);
            return escapes[i][j];
        }
        break;

    //----------- right door -------------
      case 'R':
        if (j==(M-1)) {
          escapes[i][j] = 1;
          return 1;
        }
        else{
            escapes[i][j] = check_path(doors, escapes, i, j+1, N, M);
            return escapes[i][j];
        }
        break;

      //----------- up door -------------
      case 'U':
        if (i==0){
          escapes[i][j] = 1;
          return 1;
        }
        else{
            escapes[i][j] = check_path(doors, escapes, i-1, j, N, M);
            return escapes[i][j];
        }
        break;

      //----------- down door -------------
      case 'D':
        if (i==(N-1)){
          escapes[i][j] = 1;
          return 1;
        }
        else{
            escapes[i][j] = check_path(doors, escapes, i+1, j, N, M);
            return escapes[i][j];
        }
        break;
    }
    return 0; //useless, to avoid warnings
}




int main(int argc, char** argv) { //arg = number of arguments, argv pointer array of pointers to arguments

// @@@@@@@@@@@@@@@@@@@@@@@@@@----- DECLARTATIONS -----@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
  int N,M;//number of rooms
  char **maze_doors;
  int **maze_escapes; // initializing to 0
  int escapable_rooms = 0; // number of rooms which can lead to the exit
  int non_escapable_rooms = 0;


// @@@@@@@@@@2@@@@@@@@@@@@@@@----- READING DATA -----@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

  const char * file_name = argv[1];	//name of file to open
  ifstream File(file_name); 		//open the file
  File>>N>>M;

  //creating dynamic arrays
  maze_doors = new char *[N];
  maze_escapes = new int *[N];

  for (int i=0; i<N; i++){
    maze_doors[i] = new char [M]; //initial array
    maze_escapes[i] = new int [M];//array describing escapability of each doors 
  }

  for (int i=0; i<N; i++){  //reading all rooms
    for (int j=0; j<M; j++){
      File>>maze_doors[i][j];
      maze_escapes[i][j]=0; //initializing to 0, means door has unknown escapability
    }
  }


  // @@@@@@@@@@2@@@@@@@@@@@@@@@----- SCANNING -----@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

for (int i=0;i<N;i++){
  for (int j=0;j<M;j++){
    if (maze_escapes[i][j]==0){
      int color = check_path(maze_doors, maze_escapes, i, j, N, M); //the color (1 or 3) to "paint" the path
      maze_escapes[i][j]=color; //useless, but we avoid warning
    }
    if (maze_escapes[i][j]==1) escapable_rooms++;
  }
}

  // @@@@@@@@@@@@@@@@@@@@@@@@----- PRINTING OUTPUT -----@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
  non_escapable_rooms=N*M-escapable_rooms;
  cout<<non_escapable_rooms<<"\n";

  return 0;
}

File: loop_rooms/loop_rooms.py

"""
********************************************************************************
  Project     : Programming Languages 1 - Assignment 2 - Exercise LoopRooms
  Authors     : Stylianos Kandylakis el17088, 
                Kitsos Orfanopoulos el17025
  -----------
 ProgLangI 2021, ECE-NTUA
 *******************************************************************************

>> TRANSLATING THE C++ PROBLEM:
---------------------------------
    At first we tried to "translate" our C++ code (*) to Python. That worked
    pretty much well for small testcases. Though for larger ones we exceeded
    the maximum recursion depth. Even with increasing the recursion limit 
    (sys.setrecursionlimit(5000)) the problem could not be solved, possibly
    due to memory problems. Python is not as fast as C++ and its performance 
    for recursion is very poor.

    (*): Our C++ solution was a RECURSIVE DFS at a double array. We follow
        the path of every door; if we found an exit the path was "good", else
        if we found an already visited room the path was "bad". 

>> ITERATIVE ADAPTATION:
----------------------------
    Our recursive DFS should be written in an iterative way. The Idea is to
    follow the path (do a DFS) until you find the last room, the door of which 
    leads to an exit/visited room (that will be the "paint"). Then we follow 
    again the same path from initial room to the last room, "painting" each 
    room with the same "paint". 
    "Paint" is the escapability of each room, 
        0 is unknown (initial value), 
        1 is escapable, 
        2 is visited=non escapable. 
    So in every step we follow twice each path. 
    At the end we scan our array to count escapable rooms.

>> STEPS:
----------------------------
    1. Read input and create arrays.
    2. Start following paths (DFS) from every unvisited room.
    3. Every time you finish one path, follow it again to "paint" it.
    4. Count escapable rooms and print them.
"""


import sys #for reading input from terminal


def check_next(door, escapes,i,j,N,M):
    #left door
    if (door=="L"):
        if (j==0): return (i,j,1)
        else: return (i,(j-1),escapes[i][j-1])
    #right door
    elif (door=="R"):
        if (j==(M-1)): return (i,j,1)
        else: return (i,(j+1),escapes[i][j+1])
    #right door
    elif (door=="U"):
        if (i==0): return (i,j,1)
        else: return ((i-1),j,escapes[i-1][j])
    #down door
    else:
        if (i==(N-1)): return (i,j,1)
        else: return ((i+1),j,escapes[i+1][j])

def next_indexes(i,j,door):
    #left door
    if (door=="L"):
        i,j=i,(j-1)
        #right door
    elif (door=="R"):
        i,j=i,(j+1)
        #up door
    elif (door=="U"):
        i,j=(i-1),j
        #down door
    else: i,j=(i+1),j
    return (i,j)
    



def paint_path(doors,escapes,paint,i,j,end_i,end_j):
    escapes[i][j]=paint
    # print("@@@@@@@@@@@@@@@@@@@@@@@@@@")
    # print("  Paint={}, Start=[{}][{}], End=[{}][{}]".format(paint,i,j,end_i,end_j),end=" | ")


    while((i!=end_i) or(j!=end_j)):
        
        # print("[{}][{}]".format(i,j), end=" -> ")

        door = doors[i][j]
        (i,j)=next_indexes(i,j,door)
        escapes[i][j]=paint
        
    # print()
    
def check_path(doors, escapes, i, j, N, M):
    while True:
        door = doors[i][j]
        
        # print("I am door[{}][{}]={} and I am {}".format(i,j,door,escapes[i][j]))
        
        if (escapes[i][j]!=1): escapes[i][j]=2
        (i,j,next_door) = check_next(door,escapes,i,j,N,M)
        
        # print("My next is:",next_door,"[{}][{}]".format(i,j))
      
        if (next_door!=0): break #emulate do-while
        

    return (i,j,next_door) #(end_i, end_j, paint)
  
    

        

# @@@@@@@@@@@@@@@@@@@@@@@@@@@--- MAIN ---@@@@@@@@@@@@@@@@@@@@@@@@@@@

if __name__ == "__main__":
    # Reading Data
    with open(sys.argv[1], "rt") as inputFile: # open the file, name as 
        N, M = map(int, inputFile.readline().split())
        maze_doors = [[0 for j in range(M)] for i in range(N)]
        maze_escapes = [[0 for j in range(M)] for i in range(N)]

        for i in range(N):
            maze_doors[i]= tuple(inputFile.readline())

    # Scanning
    escapable_rooms=0
    for i in range(N):
        for j in range(M):
            if (maze_escapes[i][j]==0): 
                (end_i, end_j, paint)=check_path(maze_doors,maze_escapes,i,j,N,M)
                paint_path(maze_doors, maze_escapes,paint,i,j,end_i,end_j)
            if (maze_escapes[i][j]==1): escapable_rooms+=1
    #for record in maze_escapes:
        # print(record)
    non_escapable_rooms=N*M-escapable_rooms
    print(non_escapable_rooms)